### DISCUSSION TITLE
rp2040 device mode: how to set GPIO pins used for USB data?

### DISCUSSION THREAD

**Discussion by Spongman** [hathach/tinyusb#2621] [General]
when using host mode I can use `tuh_configure` to change the GPIO pins that are used for USB data.

what's the equivalent for Device Mode ? how do i tell it which pins to use ?

**Comment by andrewleech** [comment]
The rp2040 doesn't let you change the pins used for the built in USB peripheral, there's just the set USB_DM and USB_DP pins. The alternative is to switch to using a PIO  based usb library like https://github.com/sekigon-gonnoc/Pico-PIO-USB but I can't help with thgat, I've never used it myself and it's not related to TinyUSB

**Comment by Spongman** [comment]
how come it works for TinyUSB Host mode? does _that_ use PIO?

**Comment by andrewleech** [comment]
Oh ok, according to https://docs.tinyusb.org/en/latest/reference/supported.html there is a pio_usb driver available here for rp2040, I didn't realise that was included already. Perhaps it can be used for device mode too then to select why pins. I haven't used that myself though, sorry my comments above are likely inaccurate.
